[data] store bytes spilled/restored after plan execution

Zandew opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

<!-- Please give a short summary of the change and the problem this solves. -->

Adds bytes spilled/restored to `DatasetStats` after a plan finishes execution. 

Refactors `internal_api` a bit so that we can get the actual reply instead of a prettified string.

First part of #38847.  Next steps would be to report spilling of individual blocks from ray.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

scottjlee commented on python/ray/_private/internal_api.py:
```diff
@@ -30,20 +38,16 @@ def memory_summary(
     from ray.dashboard.memory_utils import memory_summary
 
-    address = services.canonicalize_bootstrap_address_or_die(address)
+    get_state_from_address(address)
```
is this the intended use for `state`?
```suggestion
    state = get_state_from_address(address)
```

scottjlee commented on python/ray/_private/internal_api.py:
```diff
@@ -30,20 +38,16 @@ def memory_summary(
-    options = GcsClientOptions.from_gcs_address(address)
-    state._initialize_global_state(options)
     if stats_only:
         return get_store_stats(state)
```
we should update this usage of `get_store_stats` as well

scottjlee commented on python/ray/data/_internal/plan.py:
```diff
@@ -635,12 +636,18 @@ def execute(
+            # Retrieve memory-related stats from ray.
+            reply = get_memory_info_reply(get_state_from_address(None))
+            if reply.store_stats.spill_time_total_s > 0:
+                stats.bytes_spilled = int(reply.store_stats.spilled_bytes_total)
```
nit: could we name it something like `global_bytes_spilled`, to indicate that this is global (and not just ray data)?

scottjlee commented on python/ray/data/tests/test_object_gc.py:
```diff
@@ -234,6 +233,22 @@ def consume(p):
+    # The size of dataset is 500*(80*80*4)*8B, about 100MB.
+    ds = ray.data.range_tensor(500, shape=(80, 80, 4), parallelism=100).materialize()
+    
+    assert ds._get_stats_summary().bytes_spilled > 0
```
nit: can we make use of `check_no_spill(...)`/ `not check_no_spill(...)` to implement the spill check, similar to other tests in this file?

Zandew replied:
Yeah I'll add `check_no_spill` to these tests, but that function doesn't cover the new code path introduced here. This tests retrieving the stats to `DatasetStats` after plan execution. Maybe, this test should be somewhere else?

scottjlee commented on python/ray/data/tests/test_object_gc.py:
```diff
@@ -235,6 +237,28 @@ def consume(p):
+    # The size of dataset is 500*(80*80*4)*8B, about 100MB.
+    ds = ray.data.range_tensor(500, shape=(80, 80, 4), parallelism=100).materialize()
+
+    check_no_spill(ctx, ds.repeat(), spill_expected=True)
```
nit: alternatively, can we leave `check_no_spill` unmodified, and instead check that `check_no_spill` fails the assertion?
```
with pytest.raises(AssertionError):
    check_no_spill(ctx, ds.repeat())
```

bveeramani commented on python/ray/data/tests/test_object_gc.py:
```diff
@@ -235,6 +234,29 @@ def consume(p):
+    ds = ray.data.range_tensor(500, shape=(80, 80, 4), parallelism=100).materialize()
+
+    with pytest.raises(AssertionError):
+        check_no_spill(ctx, ds.repeat())
```
Out of scope for this PR, but `repeat` is deprecated and might get removed in an upcoming release. We should rewrite the tests in this module to not use `DatasetPipeline`.

bveeramani commented on python/ray/data/tests/test_object_gc.py:
```diff
@@ -235,6 +234,29 @@ def consume(p):
+    # The object store is about 90MB.
+    ctx = ray.init(object_store_memory=90e6)
+    # The size of dataset is 500*(80*80*4)*8B, about 100MB.
+    ds = ray.data.range_tensor(500, shape=(80, 80, 4), parallelism=100).materialize()
```
Won't spilling occur when you call `materialize` here?

Zandew replied:
Yes, this test checks for spills

bveeramani replied:
Ah, I guess I was wondering why we call `iter_batches` in `check_no_spill` then? Since the spilling has already occurred?

Out of scope for this PR, though. I understand that the other tests call `repeat` and this test is just following that pattern.

bveeramani commented on python/ray/_private/internal_api.py:
```diff
@@ -30,20 +39,18 @@ def memory_summary(
 
-def get_store_stats(state, node_manager_address=None, node_manager_port=None):
-    """Returns a formatted string describing memory usage in the cluster."""
+def get_memory_info_reply(state, node_manager_address=None, node_manager_port=None):
```
For my own understanding, where do I find the definition of the reply object?

Zandew replied:
I think its `ObjectStoreStats` in `node_manager.proto`

bveeramani reviewed (commented):
Overall LGTM

stephanie-wang commented on python/ray/data/_internal/stats.py:
```diff
@@ -430,6 +438,8 @@ def __repr__(self, level=0) -> str:
             f"{indent}   extra_metrics={{{extra_metrics}}},\n"
             f"{indent}   stage_stats=[{stage_stats}],\n"
             f"{indent}   iter_stats={self.iter_stats.__repr__(level+1)},\n"
+            f"{indent}   global_bytes_spilled={self.global_bytes_spilled},\n"
```
Minor suggestion to make this more human readable (convert to MB or GB).

stephanie-wang commented:
Thanks for the cleanups as well!

bveeramani reviewed (approved):
LGTM

stephanie-wang commented:
PR looks good, but I tried it out and don't see the bytes spilled when calling `print(Dataset.stats())`. We can merge this first if you want to add that in a separate PR after also getting stats from individual blocks, though.

Zandew commented:
> PR looks good, but I tried it out and don't see the bytes spilled when calling `print(Dataset.stats())`. We can merge this first if you want to add that in a separate PR after also getting stats from individual blocks, though.

Just added the spill stats to `Dataset.stats()` and updated the tests  👍

stephanie-wang commented on python/ray/data/_internal/stats.py:
```diff
@@ -407,6 +418,11 @@ def to_string(
             out += "* Extra metrics: " + str(self.extra_metrics) + "\n"
         out += str(self.iter_stats)
+        if len(self.stages_stats) > 0 and add_global_stats:
+            out += "\nGlobal memory:\n"
+            out += "* Spilled: {}MB\n".format(round(self.global_bytes_spilled / 1e6))
+            out += "* Restored: {}MB\n".format(round(self.global_bytes_restored / 1e6))
```
```suggestion
            out += "\nCluster memory:\n"
            out += "* Spilled to disk: {}MB\n".format(round(self.global_bytes_spilled / 1e6))
            out += "* Restored from disk: {}MB\n".format(round(self.global_bytes_restored / 1e6))
```

Can you also update this to only add the string if one of these values is nonzero?
